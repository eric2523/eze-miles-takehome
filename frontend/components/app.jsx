import React from "react";
import TableIndex from "./table-index";
import HistoryButtons from "./history-buttons";

const App = () => {
  return (
    <div>
      <TableIndex />
      <HistoryButtons />
      <div className="socials">
        <a
          href="https://github.com/eric2523/eze-miles-takehome"
          target="_blank"
          rel="noopener noreferrer"
        >
          <img src="https://img.icons8.com/clouds/100/000000/github.png" />
        </a>
        <a
          href="https://www.linkedin.com/in/eric-xian-2998061a6/"
          target="_blank"
          rel="noopener noreferrer"
        >
          <img src="https://img.icons8.com/clouds/100/000000/linkedin.png" />
        </a>
      </div>
    </div>
  );
};

export default App;
