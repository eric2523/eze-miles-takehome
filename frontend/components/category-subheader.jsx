import React from "react"
import { connect } from "react-redux"

const CategorySubheaderItem = ({ name }) => {
  return (
    <th className="cat-subhead">
      <h2>
        { name }
      </h2>
    </th>
  )
}

const CategorySubheader = ({ categories }) => {
  let list = [<th key={ "null-placeholder" }></th>]
  for(const id in categories.byId){
    let category = categories.byId[id]

    list.push(
      <CategorySubheaderItem key={ category.id } name={ category.name }/>
    )
  }

  return (
    <tr>
      { list }
    </tr>
  )
}

const mSTP = (state) => {
  return ({
    categories: state.entities.categories
  })
}

export default connect(mSTP, null)(CategorySubheader)