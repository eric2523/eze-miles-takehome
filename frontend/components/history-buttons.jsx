import React, { useState } from "react";
import { ActionCreators as UndoActionCreators } from 'redux-undo'
import { connect } from 'react-redux'

const HistoryButtons = ({ canUndo, canRedo, onUndo, onRedo, state}) => {
  const handleSave = () => {
    let cache = JSON.stringify(state)
    window.localStorage.setItem("cache", cache)
  }

  return (
    <div className="btn-div">
      <button data-testid="undo-btn" onClick={onUndo} disabled={!canUndo}>
        Undo
      </button>
      <button data-testid="redo-btn" onClick={onRedo} disabled={!canRedo}>
        Redo
      </button>
      <button onClick={handleSave}>
        Save
      </button>
    </div>
  );
};

const mapStateToProps = state => {
  return {
    canUndo: state.entities.categoryRewards.past.length > 0,
    canRedo: state.entities.categoryRewards.future.length > 0,
    state: state,
  }
}

const mapDispatchToProps = dispatch => {
  return {
    onUndo: () => dispatch(UndoActionCreators.undo()),
    onRedo: () => dispatch(UndoActionCreators.redo())
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(HistoryButtons)

