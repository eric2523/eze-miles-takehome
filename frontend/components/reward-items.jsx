import React from "react";

class RewardItem extends React.Component {
  constructor(props){
    super(props)
    this.handleDragStart = this.handleDragStart.bind(this)
    this.handleDragOver = this.handleDragOver.bind(this)
  }

  handleDragStart(){
    const target = event.target;
    event.dataTransfer.setData("categoryRewardId", this.props.id)
  };

  handleDragOver(){
    event.stopPropagation();
  };

  render(){
    return (
      <div
        key={this.props.id}
        className="draggables"
        onDragStart={this.handleDragStart}
        onDragOver={this.handleDragOver}
        onDrop={this.props.handleDrop(this.props.id)}
        draggable={this.props.draggable}
      >
        {this.props.name}
      </div>
    );
  }
}

export default RewardItem;
