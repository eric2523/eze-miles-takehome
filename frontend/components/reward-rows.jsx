import React from "react";
import { connect } from "react-redux";

const RewardsRow = ({ rewards, row, idx }) => {
  let data = [];
  row.forEach( rewardItem => {
    data.push(
      <td key={rewardItem.props.id}>
          { rewardItem }
      </td>
    );
  });

  // debugger
  return (
    <tr>
      <td key={ idx }>{ "R" + (idx + 1) }</td>
      { data }
    </tr>);
};

const mSTP = (state) => {
  return {
    rewards: state.entities.rewards,
  };
};

export default connect(mSTP, null)(RewardsRow);
