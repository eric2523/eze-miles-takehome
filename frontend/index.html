<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <link rel="stylesheet" href="../app/assets/stylesheets/01.css">
  <link rel="preconnect" href="https://fonts.gstatic.com">
  <link href="https://fonts.googleapis.com/css2?family=Open+Sans:wght@300;400&display=swap" rel="stylesheet">
  <title>Document</title>
</head>
<body>
  <div id="root">
    broken
  </div>
  <script src="../app/assets/javascripts/bundle.js"></script>
</body>
</html>