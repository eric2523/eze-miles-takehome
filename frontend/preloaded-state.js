export const preloadedState = {
  entities: {
    categories: {
      byId: {
        1: {
          id: 1,
          name: "Food & Drink",
        },
        2: {
          id: 2,
          name: "Automotive",
        },
        3: {
          id: 3,
          name: "Home",
        },
      },
      allIds: [1, 2, 3],
    },
    rewards: {
      byId: {
        1: {
          id: 1,
          name: "100% OFF",
        },
        2: {
          id: 2,
          name: "$0.01 OFF",
        },
        3: {
          id: 3,
          name: "BOGO",
        },
      },
      allIds: [1, 2, 3],
    },
    categoryRewards: {
      past: [],
      future: [],
      present: {
        byId: {
          1: {
            id: 1,
            categoryId: 1,
            rewardId: 1,
          },
          2: {
            id: 2,
            categoryId: 1,
            rewardId: 2,
          },
          3: {
            id: 3,
            categoryId: 2,
            rewardId: 1,
          },
          4: {
            id: 4,
            categoryId: 2,
            rewardId: 2,
          },
        },
        allIds: [1, 2, 3, 4],
      },
    },
  },
};
