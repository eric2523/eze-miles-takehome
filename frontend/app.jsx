import React from "react"
import ReactDOM from "react-dom"
import { Root } from "./components/root"
import configureStore from "./store/store"
import { receiveCategoryRewards, updateCategoryRewards, deleteCategoryRewards } from "./actions/category-rewards-actions"
import { preloadedState } from "./preloaded-state"

document.addEventListener("DOMContentLoaded", () => {
  const root = document.getElementById("root")
  let cache = JSON.parse(window.localStorage.getItem("cache"))
  const initialState = (cache) ? cache : preloadedState
  const store = configureStore(initialState)

  ReactDOM.render(<Root store={ store }/>, root)
  // testing on window stuff 
  window.store = store 
  window.receiveCategoryRewards = receiveCategoryRewards
  window.updateCategoryRewards = updateCategoryRewards
  window.deleteCategoryRewards = deleteCategoryRewards
})