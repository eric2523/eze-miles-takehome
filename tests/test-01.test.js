import React from "react";
import { render, screen } from "./test-utils";
import App from "../frontend/components/app";
import { preloadedState } from "../frontend/preloaded-state";
import "@testing-library/jest-dom";
import {
  receiveCategoryRewards,
  updateCategoryRewards,
  deleteCategoryRewards,
} from "../frontend/actions/category-rewards-actions";
import { fireEvent } from "@testing-library/dom";
import configureMockStore from "redux-mock-store";
import { Root } from "../frontend/components/root";

describe("Table", () => {
  it("Should have approriate headers and subheaders", () => {
    render(<App />, { initialState: preloadedState });

    expect(screen.getByTestId("main-table")).toBeInTheDocument();
    expect(screen.getByTestId("main-th-rew")).toHaveTextContent("rewards");
    expect(screen.getByTestId("main-th-cat")).toHaveTextContent("categories");
    expect(screen.getByTestId("main-table")).toHaveTextContent(
      "R1",
      "R2",
      "R3"
    );
  });

  describe("Table Data", () => {
    it("Should have draggable attributes", () => {
      render(<App />, { initialState: preloadedState });

      const nonEmpty = screen.getAllByTestId("row-item");
      nonEmpty.forEach((el) => {
        expect(el).toHaveAttribute("draggable");
      });

      const empty = screen.getAllByTestId("empty-row-item");
      empty.forEach((el) => {
        expect(el).toHaveAttribute("draggable");
      });
    });

    it("Empty cells should have draggable=false", () => {
      render(<App />, { initialState: preloadedState });

      const empty = screen.getAllByTestId("empty-row-item");
      empty.forEach((el) => {
        let val = el.getAttribute("draggable");
        expect(val).toEqual("false");
      });
    });

    it("Non-empty cells should have draggable=true", () => {
      render(<App />, { initialState: preloadedState });

      const nonEmpty = screen.getAllByTestId("row-item");
      nonEmpty.forEach((el) => {
        let val = el.getAttribute("draggable");
        expect(val).toEqual("true");
      });
    });
  });
});

describe("Action Creators", () => {
  let store, mockStore

  beforeEach(() => {
    mockStore = configureMockStore();
    store = mockStore(preloadedState);
  })

  it("Able to receive new categoryRewards", () => {
    store.dispatch(
      receiveCategoryRewards({
        id: 100,
        categoryId: 100,
        rewardId: 100,
      })
    );

    expect(store.getActions()).toMatchSnapshot();
    expect(store.getState().entities.categoryRewards.present.byId[100]).not
      .toBeEmpty;
  });

  it("Able to update existing categoryRewards", () => {
    store.dispatch(
      updateCategoryRewards({
        id: 1,
        categoryId: 3,
        rewardId: 3,
      })
    );

    expect(store.getActions()).toMatchSnapshot();
  });

  it("Able to delete existing categoryRewards", () => {
    store.dispatch(deleteCategoryRewards(1));

    expect(store.getActions()).toMatchSnapshot();
  });
});

describe("User Interaction", () => {
  describe("X-Button", () => {
    it("X button should remove data entry", () => {
      render(<App />, { initialState: preloadedState });
  
      const buttons = screen.getAllByTestId("x-btn");
      buttons.forEach((button) => {
        fireEvent.click(button);
      });
      let filledCells = screen.queryAllByTestId("row-item");
      expect(filledCells).toBeEmpty;
    });
  })

  describe("Undo Button", () => {
    it("Should retrieve most recent action from past", () => {
      render(<App />, { initialState: preloadedState });
      const old = screen.getByTestId("main-table").innerHTML;
      const button = screen.getAllByTestId("x-btn")[0];
      const undoBtn = screen.getByTestId("undo-btn")
  
      // click x button to create new display
      fireEvent.click(button);
      // click undo to revert back to old display
      fireEvent.click(undoBtn)
      expect(screen.getByTestId("main-table").innerHTML).toEqual(old);
    });
  })

  describe("Redo Button", () => {
    it("Should retrieve most recent action from future", () => {
      render(<App />, { initialState: preloadedState });
      const redoBtn = screen.getByTestId("redo-btn")
      const undoBtn = screen.getByTestId("undo-btn")
      const button = screen.getAllByTestId("x-btn")[0];
  
      // click x button to create new display
      fireEvent.click(button)
      // save display
      const old = screen.getByTestId("main-table").innerHTML;
      // click undo
      fireEvent.click(undoBtn)
      // click redo
      fireEvent.click(redoBtn)
      // compare displays
      expect(screen.getByTestId("main-table").innerHTML).toEqual(old);
    });
  })
});
