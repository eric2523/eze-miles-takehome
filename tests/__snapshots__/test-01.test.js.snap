// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Action Creators Able to delete existing categoryRewards 1`] = `
Array [
  Object {
    "id": 1,
    "type": "DELETE_CATEGORY_REWARD",
  },
]
`;

exports[`Action Creators Able to receive new categoryRewards 1`] = `
Array [
  Object {
    "categoryRewards": Object {
      "categoryId": 100,
      "id": 100,
      "rewardId": 100,
    },
    "type": "RECEIVE_CATEGORY_REWARD",
  },
]
`;

exports[`Action Creators Able to update existing categoryRewards 1`] = `
Array [
  Object {
    "categoryRewards": Object {
      "categoryId": 3,
      "id": 1,
      "rewardId": 3,
    },
    "type": "UPDATE_CATEGORY_REWARD",
  },
]
`;
